### PR TITLE
Fix/Optimize state store APIs

### PR DESCRIPTION
**Pull request by yuva29**
1. Write/Read/ReadAll should handle `Key not found` explicitly

2. Optimze the code to iterate only when connection is refused/etcd cluster is unavailable

Signed-off-by: Yuva Shankar <yuva29@users.noreply.github.com>

### CODE REVIEW

File: state/consulstatedriver.go [hunk 1]
```diff
@@ -70,71 +70,80 @@ func processKey(inKey string) string {
...
+			if api.IsServerError(err) || strings.Contains(err.Error(), "EOF") || strings.Contains(err.Error(), "connection refused") {
 				time.Sleep(time.Second)
+			} else {
+				return []byte{}, err
 			}
-		} else {
-			return []byte{}, err
+		} else if kv == nil {
+			return []byte{}, core.Errorf("Key not found")
+		} else if err == nil {
```
Comment (jojimt) [inline state/consulstatedriver.go]: Check for err==nil here is redundant.

File: state/consulstatedriver.go [hunk 2]
```diff
@@ -70,71 +70,80 @@ func processKey(inKey string) string {
...
+				time.Sleep(time.Second)
+				continue
+			} else {
+				return [][]byte{}, err
+			}
+		} else if kvs == nil {
+			// Consul returns success and a nil kv when a key is not found,
+			// translate it to 'Key not found' error
+			return nil, core.Errorf("Key not found")
+		} else if err == nil {
```
Comment (jojimt) [inline state/consulstatedriver.go]: Here as well.

File: state/etcdstatedriver.go [hunk 3]
```diff
@@ -105,46 +103,49 @@ func (d *EtcdStateDriver) Read(key string) ([]byte, error) {
...
+			for _, node := range resp.Node.Nodes {
+				values = append(values, []byte(node.Value))
+			}
+			return values, nil
+		} else if client.IsKeyNotFound(err) {
+			return [][]byte{}, core.Errorf("Key not found")
+		} else if err.Error() == client.ErrClusterUnavailable.Error() {
+			// Retry after a delay
+			time.Sleep(time.Second)
+		}
```
Comment (jojimt) [inline state/etcdstatedriver.go]: What happens if it's some other error?

File: state/etcdstatedriver.go [hunk 4]
```diff
@@ -105,46 +103,49 @@ func (d *EtcdStateDriver) Read(key string) ([]byte, error) {
...
+
+	for i := 0; i < maxEtcdRetries; i++ {
+		resp, err = d.KeysAPI.Get(ctx, key, &client.GetOptions{Quorum: true})
+		if err == nil {
+			return []byte(resp.Node.Value), nil
+		} else if client.IsKeyNotFound(err) {
+			return []byte{}, core.Errorf("Key not found")
+		} else if err.Error() == client.ErrClusterUnavailable.Error() {
+			// Retry after a delay
+			time.Sleep(time.Second)
```
Comment (jojimt) [inline state/etcdstatedriver.go]: What if it's some other error?

**Comment by yuva29** [comment]
@jojimt I changed the logic to handle the conditions you pointed. PTAL

File: state/etcdstatedriver.go (hunk 4 above)
Comment (jojimt) [inline state/etcdstatedriver.go]: @yuva29 Any reason why you did not retain the original logic? Are you trying to anything beyond just placing the whole logic inside a single for loop?

Comment (yuva29) [inline state/etcdstatedriver.go]: `Key not found` case wasn't handled earlier in `etcd` driver. Also, I have removed unnecessary `else` statements to align with go standards. Logic is pretty much the same except the way its coded.

Comment (jojimt) [inline state/etcdstatedriver.go]: @yuva29 Please retain the check for resp.Node != nil as discussed. Also, please eliminate unnecessary else statements in the above if chain (if there is a return, there is no need for an else. Not sure why go vet did not catch it...).

**Comment by yuva29** [comment]
@jojimt PTAL

**Pull request merged**